Should we allow polymorphic overloads with the same number of type parameters?

These two overloads are considered valid by Dotty, because their signatures are different:
```scala
class A {
  def foo[T <: Serializable](x: T): Unit = {}
  def foo[T <: Cloneable](x: T): Unit = {}
}
```
... but trying to override one of them fails, because refchecks uses `Types#matches` to find overrides which does not take signatures into account (it calls `TypeComparer#matchesType` which in turns does `matchesType(tp1.resultType, tp2.resultType.subst(tp2, tp1))`, therefore comparing `(x: T): Unit` to `(x: T): Unit`):
```scala
class B extends A {
  override def foo[T <: Serializable](x: T): Unit = {}
}
```
```scala
-- Error: try/B.scala:21:15 ----------------------------------------------------
21 |  override def foo[T <: Serializable](x: T): Unit = {}
   |               ^
   |  error overriding method foo in class A of type [T <: Cloneable](x: T): Unit;
   |    method foo of type [T <: Serializable](x: T): Unit has incompatible type
```

So the question is: should we disallow these overloads, or should we strive to support them? The answer is not clear cut to me because Scala 2 never supported them:
```scala
try/B.scala:16: error: method foo is defined twice;
  the conflicting method foo was defined at line 15:7
  def foo[S <: Cloneable](x: S): Unit = {}
      ^
```
... but Java does support them, so if we don't support them we're still in trouble when trying to override them (and in fact scala 2 cannot override them):
```java
public class A {
  public <T extends java.io.Serializable> void foo(T x) {}
  public <T extends Cloneable> void foo(T x) {}
}
```

I would be in favor of supporting them for better interoperability and because it doesn't seem too hard (require matching param signatures for Types#matches to bring it in line with Denotation#matches)

WDYT @odersky ?

--- a/compiler/src/dotty/tools/dotc/transform/OverridingPairs.scala
+++ b/compiler/src/dotty/tools/dotc/transform/OverridingPairs.scala
@@ -40,7 +40,7 @@ object OverridingPairs {
      *  relative to <base>.this do
      */
     protected def matches(sym1: Symbol, sym2: Symbol): Boolean =
-      sym1.isType || self.memberInfo(sym1).matches(self.memberInfo(sym2))
+      sym1.isType || sym1.asSeenFrom(self).matches(sym2.asSeenFrom(self))
 
     /** The symbols that can take part in an overriding pair */
     private val decls = {

--- a/tests/explicit-nulls/neg/override-java-object-arg2.scala
+++ b/tests/explicit-nulls/neg/override-java-object-arg2.scala
@@ -5,7 +5,7 @@ class Foo {
 
   def bar(): Unit = {
     val listener4 = new NotificationListener() { // error: duplicate symbol error
-      def handleNotification(n: Notification|Null, emitter: Object): Unit = {
+      def handleNotification(n: Notification|Null, emitter: Object): Unit = { // error
       }
     }
   }

--- a/tests/neg-custom-args/erased/erased-case-class.scala
+++ b/tests/neg-custom-args/erased/erased-case-class.scala
@@ -1,1 +1,1 @@
-case class Foo1(erased x: Int) // error
+case class Foo1(erased x: Int) // error // error

--- a/tests/neg/i7597.scala
+++ b/tests/neg/i7597.scala
@@ -1,13 +1,13 @@
 object Test extends App {
   def foo[S <: String]: String => Int =
-    new (String => Int) { def apply(s: S): Int = 0 } // error
+    new (String => Int) { def apply(s: S): Int = 0 } // error // error
 
   trait Fn[A, B] {
     def apply(x: A): B
   }
 
   class C[S <: String] extends Fn[String, Int] {  // error
-    def apply(s: S): Int = 0
+    def apply(s: S): Int = 0 // error
   }
 
   foo("")

--- a/tests/pos-java-interop/i9109/A.java
+++ b/tests/pos-java-interop/i9109/A.java
@@ -1,0 +1,4 @@
+public class A {
+  public <T extends java.io.Serializable> void foo(T x) {}
+  public <T extends Cloneable> void foo(T x) {}
+}

--- a/tests/pos-java-interop/i9109/B.scala
+++ b/tests/pos-java-interop/i9109/B.scala
@@ -1,0 +1,3 @@
+class B extends A {
+  override def foo[T <: Serializable](x: T): Unit = {}
+}

--- a/tests/pos/i9109.scala
+++ b/tests/pos/i9109.scala
@@ -1,0 +1,7 @@
+class A {
+  def foo[T <: Serializable](x: T): Unit = {}
+  def foo[T <: Cloneable](x: T): Unit = {}
+}
+class B extends A {
+  override def foo[T <: Serializable](x: T): Unit = {}
+}

--- a/tests/pos/zoo.scala
+++ b/tests/pos/zoo.scala
@@ -14,7 +14,7 @@ trait Animal {
 trait Cow extends Animal {
   type IsMeat = Any
   type Food <: Grass
-  def eats(food: Grass): Unit
+  def eats(food: Food): Unit
   def gets: Food
 }
 trait Lion extends Animal {

